src/atomic.rs

```diff
@@ -0,0 +1,860 @@
+use std::borrow::{Borrow, BorrowMut};
+use std::marker::PhantomData;
+use std::mem;
+use std::ops::{Deref, DerefMut};
+use std::sync::atomic::AtomicUsize;
+use std::sync::atomic::Ordering;
+
+use Scope;
+
+/// Given ordering for the success case in a compare-exchange operation, returns the strongest
+/// appropriate ordering for the failure case.
+#[inline]
+fn strongest_failure_ordering(ord: Ordering) -> Ordering {
+    use self::Ordering::*;
+    match ord {
+        Relaxed => Relaxed,
+        Release => Relaxed,
+        Acquire => Acquire,
+        AcqRel => Acquire,
+        SeqCst => SeqCst,
+        _ => SeqCst,
+    }
+}
+
+/// Panics if the pointer is not properly unaligned.
+#[inline]
+fn ensure_aligned<T>(raw: *const T) {
+    assert!(raw as usize & low_bits::<T>() == 0, "unaligned pointer");
+}
+
+/// Returns a bitmask containing the unused least significant bits of an aligned pointer to `T`.
+#[inline]
+fn low_bits<T>() -> usize {
+    (1 << mem::align_of::<T>().trailing_zeros()) - 1
+}
+
+/// Given a tagged pointer `data`, returns the same pointer, but tagged with `tag`.
+/// Panics if the tag doesn't fit into the unused bits of the pointer.
+#[inline]
+fn data_with_tag<T>(data: usize, tag: usize) -> usize {
+    let mask = low_bits::<T>();
+    assert!(tag <= mask, "tag too large to fit into the unused bits: {} > {}", tag, mask);
+    (data & !mask) | tag
+}
+
+/// An atomic pointer that can be safely shared between threads.
+///
+/// The pointer must be properly aligned. Since it is aligned, a tag can be stored into the unused
+/// least significant bits of the address.
+///
+/// Any method that loads the pointer must be passed a reference to a [`Scope`].
+///
+/// [`Scope`]: struct.Scope.html
+#[derive(Debug)]
+pub struct Atomic<T> {
+    data: AtomicUsize,
+    _marker: PhantomData<*mut T>,
+}
+
+unsafe impl<T: Send + Sync> Send for Atomic<T> {}
+unsafe impl<T: Send + Sync> Sync for Atomic<T> {}
+
+impl<T> Atomic<T> {
+    /// Returns a new atomic pointer initialized with the tagged pointer `data`.
+    fn from_data(data: usize) -> Self {
+        Atomic {
+            data: AtomicUsize::new(data),
+            _marker: PhantomData,
+        }
+    }
+
+    /// Returns a new null atomic pointer.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use crossbeam_epoch::Atomic;
+    ///
+    /// let a = Atomic::<i32>::null();
+    /// ```
+    pub fn null() -> Self {
```

Comment:
As @Vtec234 pointed out, it would be nice to implement `pub const fn null() -> Self` for nightly.